File: plugins/project-intelligence/scripts/project_intel_lib/scanner/frontend.py
from __future__ import annotations

import json
import re
from collections import Counter, defaultdict
from pathlib import Path
from typing import Any, Callable

from ..core import IncrementalScanCache, file_signature


FRONTEND_SUFFIXES = {".vue", ".tsx", ".jsx", ".ts", ".js"}
STYLE_SUFFIXES = {".scss", ".css", ".less", ".sass"}


def extract_object_argument_blocks(text: str, function_name: str) -> list[str]:
    blocks: list[str] = []
    pattern = re.compile(rf"\b{re.escape(function_name)}\s*\(\s*{{")
    for match in pattern.finditer(text):
        start = match.end() - 1
        depth = 0
        quote = ""
        escaped = False
        for idx in range(start, len(text)):
            char = text[idx]
            if quote:
                if escaped:
                    escaped = False
                elif char == "\\":
                    escaped = True
                elif char == quote:
                    quote = ""
                continue
            if char in {"'", '"', "`"}:
                quote = char
            elif char == "{":
                depth += 1
            elif char == "}":
                depth -= 1
                if depth == 0:
                    blocks.append(text[start + 1 : idx])
                    break
    return blocks


def split_top_level_items(block: str) -> list[str]:
    items: list[str] = []
    start = depth = 0
    quote = ""
    escaped = False
    for idx, char in enumerate(block):
        if quote:
            if escaped:
                escaped = False
            elif char == "\\":
                escaped = True
            elif char == quote:
                quote = ""
            continue
        if char in {"'", '"', "`"}:
            quote = char
        elif char in "([{":
            depth += 1
        elif char in ")]}":
            depth = max(0, depth - 1)
        elif char == "," and depth == 0:
            item = block[start:idx].strip()
            if item:
                items.append(item)
            start = idx + 1
    tail = block[start:].strip()
    if tail:
        items.append(tail)
    return items


def top_level_object_keys(block: str) -> list[str]:
    names: list[str] = []
    for item in split_top_level_items(block):
        match = re.match(r"(?:['\"]([^'\"]+)['\"]|([A-Za-z_$][\w$]*))\??\s*:", item)
        if match:
            names.append(match.group(1) or match.group(2))
    return names


def extract_vue_props(text: str) -> list[str]:
    names: set[str] = set()
    for block in re.findall(r"defineProps\s*<\s*{([^}]*)}", text, re.S):
        names.update(re.findall(r"\b([A-Za-z_$][\w$]*)\??\s*:", block))
    for type_name in re.findall(r"defineProps\s*<\s*([A-Za-z_$][\w$]*)\s*>", text):
        pattern = rf"(?:interface\s+{re.escape(type_name)}|type\s+{re.escape(type_name)}\s*=)\s*{{([^}}]*)}}"
        for block in re.findall(pattern, text, re.S):
            names.update(re.findall(r"\b([A-Za-z_$][\w$]*)\??\s*:", block))
    for block in extract_object_argument_blocks(text, "defineProps"):
        names.update(top_level_object_keys(block))
    return sorted(names)


def extract_emits(text: str) -> list[str]:
    names: set[str] = set()
    for block in re.findall(r"defineEmits\s*\(\s*\[([^\]]*)]", text, re.S):
        names.update(re.findall(r"['\"]([^'\"]+)['\"]", block))
    for block in re.findall(r"defineEmits\s*\(\s*{([^}]*)}", text, re.S):
        for hit in re.findall(r"(?:^|[,{\s])(?:['\"]([^'\"]+)['\"]|([A-Za-z_$][\w$]*))\s*:", block):
            names.update(part for part in hit if part)
    for block in re.findall(r"defineEmits\s*<([^>]*)>", text, re.S):
        names.update(re.findall(r"['\"]([A-Za-z0-9:_-]+)['\"]", block))
    return sorted(name for name in names if re.match(r"^[A-Za-z][A-Za-z0-9:_-]*$", name))


def extract_slots(text: str) -> list[str]:
    expanded: set[str] = {slot or "default" for slot in re.findall(r"<slot(?:\s+name=['\"]([^'\"]+)['\"])?", text)}
    for block in re.findall(r"defineSlots\s*<\s*{([^}]*)}", text, re.S):
        expanded.update(re.findall(r"\b([A-Za-z_$][\w$]*)\??\s*[:(]", block))
    return sorted(expanded)


def extract_expose(text: str) -> list[str]:
    names: set[str] = set()
    for block in extract_object_argument_blocks(text, "defineExpose"):
        for item in split_top_level_items(block):
            match = re.match(r"([A-Za-z_$][\w$]*)", item)
            if match:
                names.add(match.group(1))
    return sorted(names)


def extract_dependencies(text: str) -> list[str]:
    values = re.findall(r"(?:import\s+[^;]*?\s+from\s+|import\s*\(|require\s*\()\s*['\"]([^'\"]+)['\"]", text)
    return sorted(dict.fromkeys(values))[:50]


def component_scope(path: str) -> str:
    if path.startswith(("src/components/", "components/")):
        return "public"
    if "/components/" in path:
        return "page-local"
    if path.startswith(("src/pages/", "pages/", "app/")):
        return "page"
    return "module"


def extract_service_prefixes(text: str) -> list[dict[str, str]]:
    prefixes = []
    for name, value in re.findall(r"\bconst\s+([A-Za-z_$][\w$]*)\s*=\s*['\"]([^'\"]*/[^'\"]*)['\"]", text):
        if any(token in value for token in ("service", "openapi", "api", "adapt")):
            prefixes.append({"name": name, "value": value.rstrip("/") or value})
    return prefixes[:20]


def extract_api_endpoints(text: str) -> list[str]:
    endpoints: list[str] = []
    pattern = r"(?<![\w$])(?:\$post|\$get|\$put|\$delete|request|fetch|axios)\s*\(\s*(`([^`]+)`|['\"]([^'\"]+)['\"])"
    for match in re.finditer(pattern, text):
        raw = match.group(2) or match.group(3) or ""
        if raw:
            endpoints.append(raw)
    endpoints.extend(re.findall(r"['\"](/[^'\"]*(?:service|openapi|api|adapt)[^'\"]*)['\"]", text))
    return sorted(dict.fromkeys(endpoints))[:40]


def extract_exported_functions(text: str) -> list[str]:
    names = re.findall(r"\bexport\s+const\s+([A-Za-z_$][\w$]*)\s*=", text)
    names += re.findall(r"\bexport\s+function\s+([A-Za-z_$][\w$]*)\s*\(", text)
    return sorted(dict.fromkeys(names))[:80]


def extract_react_props(text: str) -> list[str]:
    names: set[str] = set()
    for block in re.findall(r"(?:interface|type)\s+\w*Props\w*\s*(?:=\s*)?{([^}]*)}", text, re.S):
        names.update(re.findall(r"\b([A-Za-z_$][\w$]*)\??\s*:", block))
    blocks = re.findall(r"function\s+[A-Z][A-Za-z0-9_]*\s*\(\s*{([^}]*)}", text, re.S)
    blocks += re.findall(r"=\s*\(\s*{([^}]*)}\s*\)\s*=>", text, re.S)
    for block in blocks:
        names.update(name.strip().split(":")[0].strip() for name in block.split(",") if name.strip())
    return sorted(name for name in names if re.match(r"^[A-Za-z_$][\w$]*$", name))


def route_module_info(text: str) -> dict[str, Any]:
    routes = sorted(set(re.findall(r"path\s*:\s*['\"]([^'\"]+)['\"]", text)))
    return {
        "baseUrls": sorted(set(re.findall(r"baseUrl\s*:\s*['\"]([^'\"]+)['\"]", text))),
        "routes": routes,
        "routeCount": len(routes),
        "customNavigationCount": len(re.findall(r"navigationStyle\s*:\s*['\"]custom['\"]", text)),
        "pluginProviders": sorted(set(re.findall(r"provider\s*:\s*['\"]([^'\"]+)['\"]", text))),
        "titlesSample": re.findall(r"navigationBarTitleText\s*:\s*['\"]([^'\"]+)['\"]", text)[:20],
    }


PATTERN_DEFS = {
    "table": [r"<(el-table|ProTable)\b", r"\bcolumns\s*=", r"type:\s*['\"]selection"],
    "pagination": [r"<(el-pagination|Pagination)\b", r"page(Size|Num|Index)", r"usePagination"],
    "dialog-drawer": [r"<(el-dialog|el-drawer|Drawer|nut-popup|dx-dialog)\b", r"v-model:visible", r"defineEmits|showModal"],
    "search-form": [r"<el-form\b", r"handle(Search|Reset)", r"search(Form|Params)"],
    "permission": [r"\b(permission|auth|v-permission|hasPermission)\b"],
    "export-download": [r"\b(export|download|导出)\b"],
}


def scan_frontend_file(path: Path, rel_path: str, text: str) -> dict[str, Any]:
    suffix = path.suffix.lower()
    lower = rel_path.lower()
    facts: dict[str, Any] = {key: [] for key in ("components", "hooks", "routes", "apiModules", "stores", "styles")}
    if suffix in {".vue", ".tsx", ".jsx"} and ("/components/" in lower or suffix == ".vue"):
        name = path.stem if path.stem != "index" else path.parent.name
        facts["components"].append(
            {
                "name": name,
                "path": rel_path,
                "kind": "vue" if suffix == ".vue" else "react",
                "scope": component_scope(rel_path),
                "props": extract_vue_props(text) if suffix == ".vue" else extract_react_props(text),
                "emits": extract_emits(text) if suffix == ".vue" else [],
                "slots": extract_slots(text) if suffix == ".vue" else [],
                "expose": extract_expose(text) if suffix == ".vue" else [],
                "dependencies": extract_dependencies(text),
                "level": "candidate",
            }
        )
    if re.search(r"(^|/)use[A-Z][A-Za-z0-9_]*\.(ts|tsx|js|jsx)$", rel_path):
        facts["hooks"].append({"name": path.stem, "path": rel_path, "exports": extract_exported_functions(text), "dependencies": extract_dependencies(text), "level": "candidate"})
    if "/router" in lower or "route" in path.stem.lower():
        route_info = route_module_info(text)
        if route_info["routes"]:
            facts["routes"].append({"path": rel_path, **route_info})
    if "/api/" in lower or re.search(r"\b(axios|fetch|request)\s*[.(]", text):
        facts["apiModules"].append(
            {
                "path": rel_path,
                "signals": sorted(set(re.findall(r"\b(axios|fetch|request)\b", text))),
                "wrappers": sorted(set(re.findall(r"(?<![\w$])(\$post|\$get|\$put|\$delete|axios|fetch|request)\s*\(", text))),
                "endpoints": extract_api_endpoints(text),
                "servicePrefixes": extract_service_prefixes(text),
                "exports": extract_exported_functions(text),
            }
        )
    if "/stores/" in lower and suffix in {".ts", ".js"}:
        facts["stores"].append({"path": rel_path, "definesStore": bool(re.search(r"\bdefineStore\s*\(", text)), "exports": extract_exported_functions(text)})
    if suffix in STYLE_SUFFIXES | {".vue"}:
        hardcoded = re.findall(r"#[0-9a-fA-F]{3,8}|\b\d+px\b", text)
        if hardcoded:
            facts["styles"].append({"path": rel_path, "hardcodedValuesSample": hardcoded[:20], "count": len(hardcoded)})
    facts["patterns"] = [name for name, regexes in PATTERN_DEFS.items() if sum(1 for regex in regexes if re.search(regex, text)) >= 2]
    return facts


def scan_frontend(
    root: Path,
    files: list[Path],
    *,
    read_text: Callable[[Path], str],
    rel: Callable[[Path, Path], str],
    cache: IncrementalScanCache | None = None,
) -> dict[str, Any]:
    cache = cache or IncrementalScanCache()
    groups = {key: [] for key in ("components", "hooks", "routes", "apiModules", "stores", "styles")}
    patterns: Counter[str] = Counter()
    locations: dict[str, list[str]] = defaultdict(list)
    for path in files:
        suffix = path.suffix.lower()
        if suffix not in FRONTEND_SUFFIXES | STYLE_SUFFIXES:
            continue
        rel_path = rel(root, path)
        signature = file_signature(path)
        facts = cache.get(rel_path, "frontend", signature)
        if facts is None:
            facts = scan_frontend_file(path, rel_path, read_text(path))
            cache.put(rel_path, "frontend", signature, facts)
        for key in groups:
            groups[key].extend(facts.get(key, []))
        for name in facts.get("patterns", []):
            patterns[name] += 1
            locations[name].append(rel_path)
    groups["redundancyCandidates"] = [
        {
            "type": "frontend-pattern",
            "name": name,
            "count": count,
            "locations": locations[name][:20],
            "level": "candidate",
            "recommendation": "审查重复使用是否应复用或抽取为组件/Hook；默认不阻塞。",
        }
        for name, count in patterns.items()
        if count >= 3
    ]
    groups["scanMode"] = "regex-fallback"
    return groups

File: src/scanner/frontend.ts
// Frontend scanner (phase 3.A.2), ported from scanner/frontend.py.
//
// Scans .vue/.tsx/.jsx/.ts/.js + style files for components, hooks, routes,
// apiModules, stores, styles, and redundancy candidates. Uses brace/quote-aware
// object argument parsing for defineProps/defineEmits (regex-fallback mode).

export const FRONTEND_SUFFIXES = new Set([".vue", ".tsx", ".jsx", ".ts", ".js"]);
export const STYLE_SUFFIXES = new Set([".scss", ".css", ".less", ".sass"]);

function allMatches(text: string, pattern: RegExp): RegExpExecArray[] {
  const out: RegExpExecArray[] = [];
  const re = new RegExp(pattern.source, pattern.flags.includes("g") ? pattern.flags : pattern.flags + "g");
  let m: RegExpExecArray | null;
  while ((m = re.exec(text)) !== null) {
    out.push(m);
    if (m.index === re.lastIndex) re.lastIndex++;
  }
  return out;
}

/** Extract balanced { ... } argument blocks following `functionName(`. */
export function extractObjectArgumentBlocks(text: string, functionName: string): string[] {
  const blocks: string[] = [];
  const pattern = new RegExp(`\\b${escapeRegex(functionName)}\\s*\\(\\s*{`, "g");
  for (const match of allMatches(text, pattern)) {
    const start = match.index! + match[0].length - 1;
    let depth = 0;
    let quote = "";
    let escaped = false;
    for (let idx = start; idx < text.length; idx++) {
      const char = text[idx]!;
      if (quote) {
        if (escaped) escaped = false;
        else if (char === "\\") escaped = true;
        else if (char === quote) quote = "";
        continue;
      }
      if (char === "'" || char === '"' || char === "`") {
        quote = char;
      } else if (char === "{") {
        depth++;
      } else if (char === "}") {
        depth--;
        if (depth === 0) {
          blocks.push(text.slice(start + 1, idx));
          break;
        }
      }
    }
  }
  return blocks;
}

export function splitTopLevelItems(block: string): string[] {
  const items: string[] = [];
  let start = 0;
  let depth = 0;
  let quote = "";
  let escaped = false;
  for (let idx = 0; idx < block.length; idx++) {
    const char = block[idx]!;
    if (quote) {
      if (escaped) escaped = false;
      else if (char === "\\") escaped = true;
      else if (char === quote) quote = "";
      continue;
    }
    if (char === "'" || char === '"' || char === "`") {
      quote = char;
    } else if (char === "(" || char === "[" || char === "{") {
      depth++;
    } else if (char === ")" || char === "]" || char === "}") {
      depth = Math.max(0, depth - 1);
    } else if (char === "," && depth === 0) {
      const item = block.slice(start, idx).trim();
      if (item) items.push(item);
      start = idx + 1;
    }
  }
  const tail = block.slice(start).trim();
  if (tail) items.push(tail);
  return items;
}

export function topLevelObjectKeys(block: string): string[] {
  const names: string[] = [];
  for (const item of splitTopLevelItems(block)) {
    const m = /(?:['"]([^'"]+)['"]|([A-Za-z_$][\w$]*))\??\s*:/.exec(item);
    if (m) names.push(m[1] ?? m[2]!);
  }
  return names;
}

export function extractVueProps(text: string): string[] {
  const names = new Set<string>();
  for (const m of allMatches(text, /defineProps\s*<\s*{([^}]*)}/gs)) {
    for (const p of allMatches(m[1]!, /\b([A-Za-z_$][\w$]*)\??\s*:/g)) names.add(p[1]!);
  }
  for (const m of allMatches(text, /defineProps\s*<\s*([A-Za-z_$][\w$]*)\s*>/g)) {
    const type = m[1]!;
    const pat = new RegExp(`(?:interface\\s+${escapeRegex(type)}|type\\s+${escapeRegex(type)}\\s*=)\\s*{([^}]*)}`, "gs");
    for (const block of allMatches(text, pat)) {
      for (const p of allMatches(block[1]!, /\b([A-Za-z_$][\w$]*)\??\s*:/g)) names.add(p[1]!);
    }
  }
  for (const block of extractObjectArgumentBlocks(text, "defineProps")) {
    for (const key of topLevelObjectKeys(block)) names.add(key);
  }
  return [...names].sort();
}

export function extractEmits(text: string): string[] {
  const names = new Set<string>();
  for (const m of allMatches(text, /defineEmits\s*\(\s*\[([^\]]*)\]/gs)) {
    for (const q of allMatches(m[1]!, /['"]([^'"]+)['"]/g)) names.add(q[1]!);
  }
  for (const m of allMatches(text, /defineEmits\s*\(\s*{([^}]*)}/gs)) {
    for (const q of allMatches(m[1]!, /(?:^|[,{\s])(?:['"]([^'"]+)['"]|([A-Za-z_$][\w$]*))\s*:/g)) {
      if (q[1]) names.add(q[1]);
      if (q[2]) names.add(q[2]);
    }
  }
  for (const m of allMatches(text, /defineEmits\s*<([^>]*)>/gs)) {
    for (const q of allMatches(m[1]!, /['"]([A-Za-z0-9:_-]+)['"]/g)) names.add(q[1]!);
  }
  return [...names].filter((n) => /^[A-Za-z][A-Za-z0-9:_-]*$/.test(n)).sort();
}

export function extractSlots(text: string): string[] {
  const expanded = new Set<string>();
  for (const m of allMatches(text, /<slot(?:\s+name=['"]([^'"]+)['"])?/g)) {
    expanded.add(m[1] || "default");
  }
  for (const m of allMatches(text, /defineSlots\s*<\s*{([^}]*)}/gs)) {
    for (const p of allMatches(m[1]!, /\b([A-Za-z_$][\w$]*)\??\s*[:(]/g)) expanded.add(p[1]!);
  }
  return [...expanded].sort();
}

export function extractExpose(text: string): string[] {
  const names = new Set<string>();
  for (const block of extractObjectArgumentBlocks(text, "defineExpose")) {
    for (const item of splitTopLevelItems(block)) {
      const m = /^([A-Za-z_$][\w$]*)/.exec(item);
      if (m) names.add(m[1]!);
    }
  }
  return [...names].sort();
}

export function extractDependencies(text: string): string[] {
  const values = (text.match(/(?:import\s+[^;]*?\s+from\s+|import\s*\(|require\s*\()\s*['"]([^'"]+)['"]/g) || []).map(
    (s) => s.match(/['"]([^'"]+)['"]/)![1]!
  );
  return [...new Set(values)].sort().slice(0, 50);
}

export function componentScope(path: string): string {
  if (path.startsWith("src/components/") || path.startsWith("components/")) return "public";
  if (path.includes("/components/")) return "page-local";
  if (path.startsWith("src/pages/") || path.startsWith("pages/") || path.startsWith("app/")) return "page";
  return "module";
}

export function extractServicePrefixes(text: string): { name: string; value: string }[] {
  const prefixes: { name: string; value: string }[] = [];
  for (const m of allMatches(text, /\bconst\s+([A-Za-z_$][\w$]*)\s*=\s*['"]([^'"]*\/[^'"]*)['"]/g)) {
    const value = m[2]!;
    if (["service", "openapi", "api", "adapt"].some((tok) => value.includes(tok))) {
      prefixes.push({ name: m[1]!, value: value.replace(/\/$/, "") || value });
    }
  }
  return prefixes.slice(0, 20);
}

export function extractApiEndpoints(text: string): string[] {
  const endpoints: string[] = [];
  const pattern = /(?<![\w$])(?:\$post|\$get|\$put|\$delete|request|fetch|axios)\s*\(\s*(`([^`]+)`|['"]([^'"]+)['"])/g;
  for (const m of allMatches(text, pattern)) {
    const raw = m[2] ?? m[3] ?? "";
    if (raw) endpoints.push(raw);
  }
  for (const m of allMatches(text, /['"](\/[^'"]*(?:service|openapi|api|adapt)[^'"]*)['"]/g)) {
    endpoints.push(m[1]!);
  }
  return [...new Set(endpoints)].sort().slice(0, 40);
}

export function extractExportedFunctions(text: string): string[] {
  let names = (text.match(/\bexport\s+const\s+([A-Za-z_$][\w$]*)\s*=/g) || []).map((m) => m.match(/const\s+(\w+)/)![1]!);
  names = names.concat(
    (text.match(/\bexport\s+function\s+([A-Za-z_$][\w$]*)\s*\(/g) || []).map((m) => m.match(/function\s+(\w+)/)![1]!)
  );
  return [...new Set(names)].sort().slice(0, 80);
}

export function extractReactProps(text: string): string[] {
  const names = new Set<string>();
  for (const m of allMatches(text, /(?:interface|type)\s+\w*Props\w*\s*(?:=\s*)?{([^}]*)}/gs)) {
    for (const p of allMatches(m[1]!, /\b([A-Za-z_$][\w$]*)\??\s*:/g)) names.add(p[1]!);
  }
  let blocks = (text.match(/function\s+[A-Z][A-Za-z0-9_]*\s*\(\s*{([^}]*)}/gs) || []).map((m) => m);
  blocks = blocks.concat((text.match(/=\s*\(\s*{([^}]*)}\s*\)\s*=>/gs) || []).map((m) => m));
  for (const block of blocks) {
    const inner = block.match(/{([^}]*)}/)?.[1] ?? "";
    for (const name of inner.split(",")) {
      const trimmed = name.trim().split(":")[0]!.trim();
      if (trimmed && /^[A-Za-z_$][\w$]*$/.test(trimmed)) names.add(trimmed);
    }
  }
  return [...names].sort();
}

export function routeModuleInfo(text: string): Record<string, unknown> {
  const routes = [...new Set((text.match(/path\s*:\s*['"]([^'"]+)['"]/g) || []).map((s) => s.match(/['"]([^'"]+)['"]/)![1]!))].sort();
  return {
    baseUrls: [...new Set((text.match(/baseUrl\s*:\s*['"]([^'"]+)['"]/g) || []).map((s) => s.match(/['"]([^'"]+)['"]/)![1]!))].sort(),
    routes,
    routeCount: routes.length,
    customNavigationCount: (text.match(/navigationStyle\s*:\s*['"]custom['"]/g) || []).length,
    pluginProviders: [...new Set((text.match(/provider\s*:\s*['"]([^'"]+)['"]/g) || []).map((s) => s.match(/['"]([^'"]+)['"]/)![1]!))].sort(),
    titlesSample: (text.match(/navigationBarTitleText\s*:\s*['"]([^'"]+)['"]/g) || []).map((s) => s.match(/['"]([^'"]+)['"]/)![1]!).slice(0, 20),
  };
}

export const PATTERN_DEFS: Record<string, RegExp[]> = {
  table: [/<(el-table|ProTable)\b/, /\bcolumns\s*=/, /type:\s*['"]selection/],
  pagination: [/<(el-pagination|Pagination)\b/, /page(Size|Num|Index)/, /usePagination/],
  "dialog-drawer": [/<(el-dialog|el-drawer|Drawer|nut-popup|dx-dialog)\b/, /v-model:visible/, /defineEmits|showModal/],
  "search-form": [/<el-form\b/, /handle(Search|Reset)/, /search(Form|Params)/],
  permission: [/\b(permission|auth|v-permission|hasPermission)\b/],
  "export-download": [/\b(export|download|导出)\b/],
};

export interface FrontendFacts {
  components: unknown[];
  hooks: unknown[];
  routes: unknown[];
  apiModules: unknown[];
  stores: unknown[];
  styles: unknown[];
  patterns?: string[];
  scanMode?: string;
}

function escapeRegex(s: string): string {
  return s.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
}

function extname(path: string): string {
  const i = path.lastIndexOf(".");
  return i === -1 ? "" : path.slice(i);
}

function stem(path: string): string {
  const base = path.split("/").pop() ?? path;
  const i = base.lastIndexOf(".");
  return i === -1 ? base : base.slice(0, i);
}

/** Scan a single frontend file (mirrors scan_frontend_file). */
export function scanFrontendFile(relPath: string, text: string): FrontendFacts {
  const suffix = extname(relPath).toLowerCase();
  const lower = relPath.toLowerCase();
  const facts: FrontendFacts = { components: [], hooks: [], routes: [], apiModules: [], stores: [], styles: [] };
  if ((suffix === ".vue" || suffix === ".tsx" || suffix === ".jsx") && (lower.includes("/components/") || suffix === ".vue")) {
    const name = stem(relPath) !== "index" ? stem(relPath) : (relPath.split("/").slice(-2)[0] ?? stem(relPath));
    facts.components.push({
      name,
      path: relPath,
      kind: suffix === ".vue" ? "vue" : "react",
      scope: componentScope(relPath),
      props: suffix === ".vue" ? extractVueProps(text) : extractReactProps(text),
      emits: suffix === ".vue" ? extractEmits(text) : [],
      slots: suffix === ".vue" ? extractSlots(text) : [],
      expose: suffix === ".vue" ? extractExpose(text) : [],
      dependencies: extractDependencies(text),
      level: "candidate",
    });
  }
  if (/(^|\/)use[A-Z][A-Za-z0-9_]*\.(ts|tsx|js|jsx)$/.test(relPath)) {
    facts.hooks.push({
      name: stem(relPath),
      path: relPath,
      exports: extractExportedFunctions(text),
      dependencies: extractDependencies(text),
      level: "candidate",
    });
  }
  if (lower.includes("/router") || stem(relPath).toLowerCase().includes("route")) {
    const info = routeModuleInfo(text);
    if ((info.routes as string[]).length) facts.routes.push({ path: relPath, ...info });
  }
  if (lower.includes("/api/") || /\b(axios|fetch|request)\s*[.(]/.test(text)) {
    facts.apiModules.push({
      path: relPath,
      signals: [...new Set((text.match(/\b(axios|fetch|request)\b/g) || []) as string[])].sort(),
      wrappers: [...new Set((text.match(/(?<![\w$])(\$post|\$get|\$put|\$delete|axios|fetch|request)\s*\(/g) || []).map((s) => s.match(/(\$\w+|\w+)/)![1]!))].sort(),
      endpoints: extractApiEndpoints(text),
      servicePrefixes: extractServicePrefixes(text),
      exports: extractExportedFunctions(text),
    });
  }
  if (lower.includes("/stores/") && (suffix === ".ts" || suffix === ".js")) {
    facts.stores.push({
      path: relPath,
      definesStore: /\bdefineStore\s*\(/.test(text),
      exports: extractExportedFunctions(text),
    });
  }
  if (STYLE_SUFFIXES.has(suffix) || suffix === ".vue") {
    const hardcoded = text.match(/#[0-9a-fA-F]{3,8}|\b\d+px\b/g) || [];
    if (hardcoded.length) facts.styles.push({ path: relPath, hardcodedValuesSample: hardcoded.slice(0, 20), count: hardcoded.length });
  }
  facts.patterns = Object.entries(PATTERN_DEFS)
    .filter(([, regexes]) => regexes.filter((re) => re.test(text)).length >= 2)
    .map(([name]) => name);
  return facts;
}

export interface ScanFrontendResult extends Omit<FrontendFacts, "scanMode"> {
  redundancyCandidates: unknown[];
  scanMode: string;
}

/** Scan a list of frontend files, aggregating facts + redundancy candidates. */
export function scanFrontend(
  files: { path: string; text?: string; readText?: () => string; signature?: string }[],
  cache?: import("./core.js").IncrementalScanCache
): ScanFrontendResult {
  const groups: Record<string, unknown[]> = { components: [], hooks: [], routes: [], apiModules: [], stores: [], styles: [] };
  const patterns = new Map<string, number>();
  const locations = new Map<string, string[]>();
  for (const file of files) {
    const { path } = file;
    const suffix = extname(path).toLowerCase();
    if (!FRONTEND_SUFFIXES.has(suffix) && !STYLE_SUFFIXES.has(suffix)) continue;
    let facts = file.signature && cache
      ? cache.get(path, "frontend", file.signature) as FrontendFacts | null
      : null;
    if (!facts) {
      const text = file.text ?? file.readText?.() ?? "";
      facts = scanFrontendFile(path, text);
      if (file.signature && cache) cache.put(path, "frontend", file.signature, facts);
    }
    for (const key of ["components", "hooks", "routes", "apiModules", "stores", "styles"] as const) {
      (groups[key] as unknown[]).push(...(facts[key] as unknown[]));
    }
    for (const name of facts.patterns ?? []) {
      patterns.set(name, (patterns.get(name) ?? 0) + 1);
      const arr = locations.get(name) ?? [];
      arr.push(path);
      locations.set(name, arr);
    }
  }
  const redundancyCandidates = [...patterns.entries()]
    .filter(([, count]) => count >= 3)
    .map(([name, count]) => ({
      type: "frontend-pattern",
      name,
      count,
      locations: (locations.get(name) ?? []).slice(0, 20),
      level: "candidate",
      recommendation: "审查重复使用是否应复用或抽取为组件/Hook；默认不阻塞。",
    }));
  return { ...groups, redundancyCandidates, scanMode: "regex-fallback" } as ScanFrontendResult;
}
